File: Cpp/2-Sum.cpp
class Solution {
public:
    /*
     * @param numbers : An array of Integer
     * @param target : target = numbers[index1] + numbers[index2]
     * @return : [index1+1, index2+1] (index1 < index2)
     */
    vector<int> twoSum(vector<int> &nums, int target) {
        // write your code here
        //O(N) time O(N) space
        unordered_map<int, int> hash;
        vector<int> res, cnt;
        for (int i = 0; i < nums.size(); i ++) {
            if (nums[i] * 2 == target) {
                cnt.push_back(i + 1);
            } else {
                hash[nums[i]] = i + 1;
            }
        }
        if (cnt.size() >= 2) {
            res.push_back(cnt[0]);
            res.push_back(cnt[1]);
            return res;
        }
        for (int i = 0; i < nums.size(); i ++) {
            if (hash.find(target - nums[i]) != hash.end()) {
                int temp = hash[target - nums[i]];
                res.push_back(min(temp, i + 1));
                res.push_back(max(temp, i + 1));
                return res;
            }
        }
        return res;
    }
};

File: Python/2-Sum.py
class Solution:
    """
    @param numbers : An array of Integer
    @param target : target = numbers[index1] + numbers[index2]
    @return : [index1 + 1, index2 + 1] (index1 < index2)
    """
    def twoSum(self, numbers, target):
        # write your code here
        dict = {}
        for i in range(len(numbers)):
            if numbers[i] not in dict:
                dict[numbers[i]] = []
            dict[numbers[i]].append(i + 1)
        seq = []
        for item in numbers:
            if item * 2 == target:
                if len(dict[item]) >= 2:
                    seq.append(min(dict[item][0], dict[item][1]))
                    seq.append(max(dict[item][0], dict[item][1]))
                    return seq
            elif target - item in dict:
                seq.append(min(dict[item][0], dict[target - item][0]))
                seq.append(max(dict[item][0], dict[target - item][0]))
                return seq
        return seq
